vowpalwabbit/core/src/reductions/binary.cc

```diff
@@ -52,16 +52,30 @@ void predict_or_learn(binary_data& data, VW::LEARNER::single_learner& base, VW::
   }
 }
 
-VW::LEARNER::base_learner* VW::reductions::binary_setup(setup_base_i& stack_builder)
+struct options_binary_v1
 {
-  options_i& options = *stack_builder.get_options();
-
   bool binary = false;
+};
+
+std::unique_ptr<options_binary_v1> get_binary_options_instance(
+    const VW::workspace&, VW::io::logger&, options_i& options)
+{
+  auto binary_opts = VW::make_unique<options_binary_v1>();
```

Comment:
IMO we should allocate the options object after calling `add_parse_and_check_necessary` and then initialize and return the object all in one go. We create and delete a lot of options objects when we dont need to.

Reply:
Not sure what you're suggesting here. `add_parse_and_check_necessary` has to be called after the option group is defined and the options are assigned to variables. Where would we assign the options if not the options_config object?